##e : Building a regression model using features like Ram, Inches, CPU_freq, PrimaryStorage, and GPU_company to predict the Price_euros of a laptop. Need to suggest the best regression model


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer


In [ ]:
data=pd.read_csv('/content/laptop_prices.csv')

In [ ]:
print(data.head())

  Company      Product   TypeName  Inches  Ram     OS  Weight  Price_euros  \
0   Apple  MacBook Pro  Ultrabook    13.3    8  macOS    1.37      1339.69   
1   Apple  Macbook Air  Ultrabook    13.3    8  macOS    1.34       898.94   
2      HP       250 G6   Notebook    15.6    8  No OS    1.86       575.00   
3   Apple  MacBook Pro  Ultrabook    15.4   16  macOS    1.83      2537.45   
4   Apple  MacBook Pro  Ultrabook    13.3    8  macOS    1.37      1803.60   

     Screen  ScreenW  ...  RetinaDisplay CPU_company CPU_freq      CPU_model  \
0  Standard     2560  ...            Yes       Intel      2.3        Core i5   
1  Standard     1440  ...             No       Intel      1.8        Core i5   
2   Full HD     1920  ...             No       Intel      2.5  Core i5 7200U   
3  Standard     2880  ...            Yes       Intel      2.7        Core i7   
4  Standard     2560  ...            Yes       Intel      3.1        Core i5   

  PrimaryStorage  SecondaryStorage PrimaryStorageT

In [ ]:
print(data.isnull().sum())

Company                   0
Product                   0
TypeName                 12
Inches                   12
Ram                       0
OS                        0
Weight                    0
Price_euros               0
Screen                    0
ScreenW                   0
ScreenH                   0
Touchscreen             589
IPSpanel                  0
RetinaDisplay             0
CPU_company               0
CPU_freq                  0
CPU_model                 5
PrimaryStorage            0
SecondaryStorage          0
PrimaryStorageType        0
SecondaryStorageType      0
GPU_company               0
GPU_model                 0
dtype: int64


In [ ]:
avg_Inches = data['Inches'].mean()
data['Inches'].fillna(avg_Inches, inplace=True)

<ipython-input-5-5e519fb618ed>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Inches'].fillna(avg_Inches, inplace=True)


In [ ]:
df = data.drop(['Product','TypeName','Screen','ScreenW','Touchscreen','RetinaDisplay','CPU_company',
          'CPU_model', 'SecondaryStorage', 'SecondaryStorageType', 'OS','GPU_model'], axis=1)

In [ ]:
print(df.head())

  Company  Inches  Ram  Weight  Price_euros  ScreenH IPSpanel  CPU_freq  \
0   Apple    13.3    8    1.37      1339.69     1600      Yes       2.3   
1   Apple    13.3    8    1.34       898.94      900       No       1.8   
2      HP    15.6    8    1.86       575.00     1080       No       2.5   
3   Apple    15.4   16    1.83      2537.45     1800      Yes       2.7   
4   Apple    13.3    8    1.37      1803.60     1600      Yes       3.1   

   PrimaryStorage PrimaryStorageType GPU_company  
0             128                SSD       Intel  
1             128      Flash Storage       Intel  
2             256                SSD       Intel  
3             512                SSD         AMD  
4             256                SSD       Intel  


In [ ]:
print(df.isnull().sum())

Company               0
Inches                0
Ram                   0
Weight                0
Price_euros           0
ScreenH               0
IPSpanel              0
CPU_freq              0
PrimaryStorage        0
PrimaryStorageType    0
GPU_company           0
dtype: int64


In [ ]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['Company'] = le.fit_transform(df['Company'])
df['GPU_company'] = le.fit_transform(df['GPU_company'])
df['PrimaryStorageType'] = le.fit_transform(df['PrimaryStorageType'])
df['IPSpanel'] = le.fit_transform(df['IPSpanel'])

In [ ]:
print(df.head())

   Company  Inches  Ram  Weight  Price_euros  ScreenH  IPSpanel  CPU_freq  \
0        1    13.3    8    1.37      1339.69     1600         1       2.3   
1        1    13.3    8    1.34       898.94      900         0       1.8   
2        7    15.6    8    1.86       575.00     1080         0       2.5   
3        1    15.4   16    1.83      2537.45     1800         1       2.7   
4        1    13.3    8    1.37      1803.60     1600         1       3.1   

   PrimaryStorage  PrimaryStorageType  GPU_company  
0             128                   3            2  
1             128                   0            2  
2             256                   3            2  
3             512                   3            0  
4             256                   3            2  


In [ ]:
x=df.drop('Price_euros',axis=1)
y=df['Price_euros']

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 2)

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(x_train)

x_train_scaled = scaler.transform(x_train)
x_test_scaled = scaler.transform(x_test)

In [ ]:
from sklearn.linear_model import LinearRegression
model_lr = LinearRegression()
model_lr.fit(x_train_scaled, y_train)

LinearRegression()

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
y_pred_lr = model_lr.predict(x_test_scaled)
mse_lr = mean_squared_error(y_test, y_pred_lr)
r2_lr = r2_score(y_test, y_pred_lr)
print("Linear Regression - Mean Squared Error:", mse_lr)
print("Linear Regression - R2 Score:", r2_lr)


Linear Regression - Mean Squared Error: 171559.3334474409
Linear Regression - R2 Score: 0.6594523944888941


In [ ]:
from sklearn.ensemble import RandomForestRegressor
model_rf = RandomForestRegressor(n_estimators=100, random_state=42)
model_rf.fit(x_train_scaled, y_train)

RandomForestRegressor(random_state=42)

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
y_pred_rf = model_rf.predict(x_test_scaled)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)
print("Random Forest - Mean Squared Error:", mse_rf)
print("Random Forest - R2 Score:", r2_rf)


Random Forest - Mean Squared Error: 121477.61419098191
Random Forest - R2 Score: 0.7588652869847233


In [ ]:
from sklearn.tree import DecisionTreeRegressor
model_dt = DecisionTreeRegressor(random_state=42)
model_dt.fit(x_train_scaled, y_train)

DecisionTreeRegressor(random_state=42)

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
y_pred_dt = model_rf.predict(x_test_scaled)
mse_dt = mean_squared_error(y_test, y_pred_dt)
r2_dt = r2_score(y_test, y_pred_dt)
print("Decision Tree - Mean Squared Error:", mse_dt)
print("Decision Tree - R2 Score:", r2_dt)

Decision Tree - Mean Squared Error: 121477.61419098191
Decision Tree - R2 Score: 0.7588652869847233


In [ ]:
from sklearn.svm import SVR
model_svm = SVR()
model_svm.fit(x_train_scaled, y_train)

SVR()

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
y_pred_svm = model_rf.predict(x_test_scaled)
mse_svm = mean_squared_error(y_test, y_pred_svm)
r2_svm = r2_score(y_test, y_pred_svm)
print("Support Vector Machine - Mean Squared Error:", mse_svm)
print("Support Vector Machine - R2 Score:", r2_svm)

Support Vector Machine - Mean Squared Error: 121477.61419098191
Support Vector Machine - R2 Score: 0.7588652869847233


In [ ]:
from sklearn.preprocessing import PolynomialFeatures
x_train_poly = PolynomialFeatures(degree=2).fit_transform(x_train_scaled)
x_test_poly = PolynomialFeatures(degree=2).fit_transform(x_test_scaled)
regressor = LinearRegression()
regressor.fit(x_train_poly, y_train)



LinearRegression()

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
y_pred_poly = regressor.predict(x_test_poly)
mse_poly = mean_squared_error(y_test, y_pred_poly)
r2_poly = r2_score(y_test, y_pred_poly)
print("Polynomial Regression - Mean Squared Error:", mse_poly)
print("Polynomial Regression - R2 Score:", r2_poly)

Polynomial Regression - Mean Squared Error: 148702.38385162465
Polynomial Regression - R2 Score: 0.704823749679709


Poly. Reg.: MSE 14870,R2 Score: 0.7048
SVM - MSE: 121477, R2 Score: 0.7588
Deci. Tree - MSE: 121477,
          R2 Sc:07588                               
Random Forest - MSE: 121477.61, R2 Score: 0.7588
Linear Regression - MSE: 171559.33,R2 Score: 0.6594
##If minimizing error (MSE) is our priority, Polynomial Regression is the best.If maximizing the model's ability to explain the variance (R²) is more important, then SVM, Decision Tree, or Random Forest are better choices.

f:Creating a classification model to predict the laptop type (TypeName) based on features such as Inches, Ram, PrimaryStorage, and Weight. Identify which features are the most important in your classification model. Suggest the best classification model

In [ ]:
data_class= pd.read_csv('/content/laptop_prices.csv')

In [ ]:
data_class.drop(['Product','Screen','ScreenW','Touchscreen','RetinaDisplay','CPU_company',
         'SecondaryStorage', 'SecondaryStorageType', 'OS','GPU_model'], axis=1,inplace=True)

In [ ]:
data_class.head()

Company   TypeName  Inches  Ram  Weight  Price_euros  ScreenH IPSpanel  \
0   Apple  Ultrabook    13.3    8    1.37      1339.69     1600      Yes   
1   Apple  Ultrabook    13.3    8    1.34       898.94      900       No   
2      HP   Notebook    15.6    8    1.86       575.00     1080       No   
3   Apple  Ultrabook    15.4   16    1.83      2537.45     1800      Yes   
4   Apple  Ultrabook    13.3    8    1.37      1803.60     1600      Yes   

   CPU_freq      CPU_model  PrimaryStorage PrimaryStorageType GPU_company  
0       2.3        Core i5             128                SSD       Intel  
1       1.8        Core i5             128      Flash Storage       Intel  
2       2.5  Core i5 7200U             256                SSD       Intel  
3       2.7        Core i7             512                SSD         AMD  
4       3.1        Core i5             256                SSD       Intel

In [ ]:
data_class.isnull().sum()

Company                0
TypeName              12
Inches                12
Ram                    0
Weight                 0
Price_euros            0
ScreenH                0
IPSpanel               0
CPU_freq               0
CPU_model              5
PrimaryStorage         0
PrimaryStorageType     0
GPU_company            0
dtype: int64

In [ ]:
for col in data_class.columns:
  if data_class[col].dtype == 'object':
    data_class[col].fillna(data_class[col].mode()[0], inplace=True)
  elif data_class[col].dtype != 'object':
    data_class[col].fillna(data_class[col].mean(), inplace=True)




<ipython-input-37-17269444e2c1>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data_class[col].fillna(data_class[col].mode()[0], inplace=True)
<ipython-input-37-17269444e2c1>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=T

In [ ]:
print(data_class.isnull().sum())

Company               0
TypeName              0
Inches                0
Ram                   0
Weight                0
Price_euros           0
ScreenH               0
IPSpanel              0
CPU_freq              0
CPU_model             0
PrimaryStorage        0
PrimaryStorageType    0
GPU_company           0
dtype: int64


In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
for col in data_class.columns:
  if data_class[col].dtype == 'object':
    data_class[col] = le.fit_transform(data_class[col])

In [ ]:
data_class.head()

Company  TypeName  Inches  Ram  Weight  Price_euros  ScreenH  IPSpanel  \
0        1         4    13.3    8    1.37      1339.69     1600         1   
1        1         4    13.3    8    1.34       898.94      900         0   
2        7         3    15.6    8    1.86       575.00     1080         0   
3        1         4    15.4   16    1.83      2537.45     1800         1   
4        1         4    13.3    8    1.37      1803.60     1600         1   

   CPU_freq  CPU_model  PrimaryStorage  PrimaryStorageType  GPU_company  
0       2.3         40             128                   3            2  
1       1.8         40             128                   0            2  
2       2.5         46             256                   3            2  
3       2.7         54             512                   3            0  
4       3.1         40             256                   3            2

In [ ]:
from re import X
X = data_class.drop('TypeName', axis=1)
Y = data_class['TypeName']

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [ ]:
from sklearn.tree import DecisionTreeClassifier
dt_model = DecisionTreeClassifier(min_samples_leaf=5)
dt_model.fit(X_train_scaled, Y_train)


DecisionTreeClassifier(min_samples_leaf=5)

In [ ]:
from sklearn.metrics import accuracy_score, classification_report
Y_pred_dt = dt_model.predict(X_test_scaled)
accuracy_dt = accuracy_score(Y_test, Y_pred_dt)
report_dt = classification_report(Y_test, Y_pred_dt)
print("Decision Tree Accuracy:", accuracy_dt)
print("Decision Tree Classification Report:\n", report_dt)

Decision Tree Accuracy: 0.7568627450980392
Decision Tree Classification Report:
               precision    recall  f1-score   support

           0       0.45      0.32      0.38        28
           1       0.83      0.88      0.85        43
           2       0.50      0.25      0.33         4
           3       0.81      0.87      0.84       141
           4       0.70      0.64      0.67        33
           5       0.29      0.33      0.31         6

    accuracy                           0.76       255
   macro avg       0.60      0.55      0.56       255
weighted avg       0.74      0.76      0.75       255



In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=100, min_samples_leaf=5)
rf_model.fit(X_train_scaled, Y_train)


RandomForestClassifier(min_samples_leaf=5)

In [ ]:
from sklearn.metrics import accuracy_score, classification_report
Y_pred_rf = rf_model.predict(X_test_scaled)
accuracy_rf = accuracy_score(Y_test, Y_pred_rf)
report_rf = classification_report(Y_test, Y_pred_rf)
print("Random Forest Accuracy:", accuracy_rf)
print("Random Forest Classification Report:\n", report_rf)

Random Forest Accuracy: 0.7843137254901961
Random Forest Classification Report:
               precision    recall  f1-score   support

           0       0.83      0.18      0.29        28
           1       0.81      0.91      0.86        43
           2       1.00      0.50      0.67         4
           3       0.82      0.91      0.87       141
           4       0.60      0.76      0.67        33
           5       0.00      0.00      0.00         6

    accuracy                           0.78       255
   macro avg       0.68      0.54      0.56       255
weighted avg       0.78      0.78      0.75       255



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:
from sklearn.svm import SVC
svm_model = SVC()
svm_model.fit(X_train_scaled, Y_train)


SVC()

In [ ]:
from sklearn.metrics import accuracy_score, classification_report
Y_pred_svm = svm_model.predict(X_test_scaled)
accuracy_svm = accuracy_score(Y_test, Y_pred_svm)
report_svm = classification_report(Y_test, Y_pred_svm)
print("SVM Accuracy:", accuracy_svm)
print("SVM Classification Report:\n", report_svm)

SVM Accuracy: 0.7333333333333333
SVM Classification Report:
               precision    recall  f1-score   support

           0       0.33      0.04      0.06        28
           1       0.70      0.81      0.75        43
           2       1.00      0.50      0.67         4
           3       0.78      0.90      0.84       141
           4       0.59      0.67      0.63        33
           5       0.00      0.00      0.00         6

    accuracy                           0.73       255
   macro avg       0.57      0.49      0.49       255
weighted avg       0.68      0.73      0.69       255



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


## Random forest Classifier is the best classifier model